Moblin/HaishinKit/Media/VideoUnit.swift

```diff
@@ -37,68 +37,83 @@ private struct FaceDetectionsCompletion {
-            }
-            // Just for sanity. Should depend on FPS and latency.
-            if sampleBuffers.count > 200 {
-                // logger.info("Over 200 frames buffered. Dropping oldest frame.")
-                sampleBuffer = replaceSampleBuffer
-                sampleBuffers.remove(at: 0)
-                continue
-            }
-            let presentationTimeStamp = replaceSampleBuffer.presentationTimeStamp.seconds
-            if firstPresentationTimeStamp.isNaN {
-                firstPresentationTimeStamp = realPresentationTimeStamp - presentationTimeStamp
-            }
-            if firstPresentationTimeStamp + presentationTimeStamp + latency > realPresentationTimeStamp {
-                break
-            }
-            sampleBuffer = replaceSampleBuffer
-            sampleBuffers.remove(at: 0)
+    func appendSampleBuffer(_ sampleBuffer: CMSampleBuffer) {
+        sampleBufferQueue.append(sampleBuffer)
+        sampleBufferQueue.sort { sampleBuffer1, sampleBuffer2 in
+            sampleBuffer1.presentationTimeStamp < sampleBuffer2.presentationTimeStamp
+        }
+        logger.info("sampleBufferQueue Count: \(sampleBufferQueue.count)")
+        if isInitialBufferingComplete == false {
+            logger.info("Starting ReplaceVideo buffering.")
+            startInitialBufferingTimer()
         }
-        currentSampleBuffer = sampleBuffer
     }
 
-    func getSampleBuffer(_ realSampleBuffer: CMSampleBuffer) -> CMSampleBuffer? {
-        if let currentSampleBuffer {
-            return makeSampleBuffer(
-                realSampleBuffer: realSampleBuffer,
-                replaceSampleBuffer: currentSampleBuffer
-            )
-        } else {
-            return nil
+    private func startInitialBufferingTimer() {
+        isInitialBufferingComplete = true
+        DispatchQueue.main.asyncAfter(deadline: .now() + latency) {
+            self.startOutput()
         }
     }
 
-    private func makeSampleBuffer(realSampleBuffer: CMSampleBuffer,
-                                  replaceSampleBuffer: CMSampleBuffer) -> CMSampleBuffer?
-    {
-        guard let sampleBuffer = CMSampleBuffer.create(replaceSampleBuffer.imageBuffer!,
-                                                       replaceSampleBuffer.formatDescription!,
-                                                       realSampleBuffer.duration,
-                                                       realSampleBuffer.presentationTimeStamp,
-                                                       realSampleBuffer.decodeTimeStamp)
-        else {
-            return nil
+    func startOutput() {
+        let frameInterval = 1 / frameRate
+        outputTimer = DispatchSource.makeTimerSource(queue: lockQueue)
+        outputTimer!.schedule(deadline: .now(), repeating: frameInterval)
+        outputTimer!.setEventHandler { [weak self] in
+            self?.outputSampleBuffer()
+        }
+        outputTimer!.activate()
```

Comment:
This outputs frames at the configured RTMP server stream FPS, which makes sense. The problem is that this output rate makes the live stream (out of Moblin) the same FPS, not the FPS configured in Settings -> Streams -> My stream -> Video -> FPS. The face effect expects frames to arrive at live stream FPS, so its timing will be incorrect. I haven't made up my mind if it is the face effect implementation that should be changed to fix this. I do think so, I think =)

Long term multi camera may be a thing, and that of course means merging multiple video feeds, with potentially different FPS:s, into a single feed with one FPS. But that's too far into the future to take into account now.

Reply:
As a builtin camera is used for output FPS it will follow that. The RTMP frames will be duplicated of dropped if needed.